library/general/src/lib/y2issues/presenter.rb

```diff
@@ -0,0 +1,125 @@
+# find current contact information at www.suse.com.
+
+require "yast"
+Yast.import "HTML"
+Yast.import "RichText"
+
+module Y2Issues
+  # This class converts a list of issues into a message for users
+  #
+  # @todo Separate by severity, group items, etc.
+  class Presenter
+    include Yast::I18n
+
+    # @return [List] List of issues to present
+    attr_reader :issues
+
+    # @param issues [List] Issues list
+    def initialize(issues)
+      textdomain "base"
+      @issues = issues
+    end
+
+    # Return the text to be shown to the user regarding the list of issues
+    #
+    # @return [String] Plain text
+    def to_plain
+      Yast::RichText.Rich2Plain(to_html)
+    end
+
+    # Return the HTML representation of a list of issues
+    #
+    # @return [String] HTML representing the list of issues
+    def to_html
+      error, non_error = issues.partition(&:error?)
+      parts = []
+      parts << error_text(error) unless error.empty?
+      parts << warning_text(non_error) unless non_error.empty?
+
+      parts.join
+    end
+
+  private
+
+    # Return warning message with a list of issues
+    #
+    # @param issues [Array<Installation::AutoinstIssues::Issue>] Array containing issues
+    # @return [String] Message
+    def warning_text(issues)
+      Yast::HTML.Para(
+        _("Minor issues were detected:")
+      ) + issues_list_content(issues)
+    end
+
+    # Return error message with a list of issues
+    #
+    # @param issues [Array<Installation::AutoinstIssues::Issue>] Array containing issues
+    # @return [String] Message
+    def error_text(issues)
+      Yast::HTML.Para(
+        _("Important issues were detected:")
+      ) + issues_list_content(issues)
+    end
+
+    # Return an HTML representation for a list of issues
+    #
+    # The issues are grouped by the location where they were detected. General issues (with no
+    # specific location) are listed first.
+    #
+    # @return [String] Issues list content
+    #
+    # @see issues_by_location
+    def issues_list_content(issues)
+      all_issues = []
+      issues_map = issues_by_location(issues)
+
+      if issues_map[:nolocation]
+        all_issues += issues_map[:nolocation].map(&:message)
+        issues_map.delete(:nolocation)
+      end
+
+      issues_map.each do |group, items|
+        messages = Yast::HTML.List(
+          items.map { |i| "#{i.location.id}: #{i.message}" }
+        )
+        path = group.split(":").last
+        all_issues << "#{path}:#{messages}"
+      end
+
+      Yast::HTML.List(all_issues)
+    end
+
+    # Return issues grouped by location where they were found
+    #
+    # @return [Hash<(#parent,#section_name),Installation::AutoinstIssues::Issue>]
+    #         Issues grouped by AutoYaST profile section
+    def issues_by_location(issues)
+      issues.each_with_object({}) do |issue, all|
+        group = if issue.location
+          "#{issue.location.type}:#{issue.location.path}"
```

Comment:
why not ID? I expect that this kind of string output is part of location itself.

Reply:
Because we use `type + path` to group errors. For instance, let's consider that we have errors:

* Error 1: `file:/etc/sysconfig/network/ifcfg-eth0:BOOTPROTO`
* Error 2: `file:/etc/sysconfig/network/ifcfg-eth0:STARTMODE`
* Error 3: `file:/etc/sysconfig/network/ifcfg-eth1:BOOTPROTO` (not it is in a different file).

Then, the presenters will group *eth0* and *eth1* errors. See [this example](https://user-images.githubusercontent.com/15836/115024267-c23e5200-9eb7-11eb-9b97-d580dfe43359.png).

Reply:
then I expect that issue has method `group` that can be called there. In general I do not like this string manipulation from other class.

Reply:
Initially, I had a method `#group`, but how the things are grouped (or not) are decided by the presenter. So I decided to keep this logic in the presenter.